### PR TITLE
Retry round tripper

### PR DESCRIPTION
**Pull request by SimonRichardson**

### CODE REVIEW

**Comment by SimonRichardson** [comment]
!!build!!

File: middleware.go [hunk 1]
```diff
@@ -131,3 +138,114 @@ func (lr roundTripRecorder) RoundTrip(req *http.Request) (*http.Response, error)
...
+		// a potential Retry-After header.
+		return res, true, nil
+	default:
+		// Don't handle any of the following status codes.
+		return res, false, nil
+	}
+}
+
+// defaultBackoff attempts to workout a good backoff strategy based on the
+// backoff policy or the status code from the response. The
```
Comment (hmlanigan) [inline middleware.go]: "The "?

File: middleware.go [hunk 2]
```diff
@@ -131,3 +138,114 @@ func (lr roundTripRecorder) RoundTrip(req *http.Request) (*http.Response, error)
...
+	}
+}
+
+type retryableErr struct{}
+
+func (retryableErr) Error() string {
+	return "retryable error"
+}
+
+func (m retryMiddleware) RoundTrip(req *http.Request) (*http.Response, error) {
```
Comment (hmlanigan) [inline middleware.go:189]: Method comment header needed.

File: middleware.go [hunk 3]
```diff
@@ -131,3 +138,114 @@ func (lr roundTripRecorder) RoundTrip(req *http.Request) (*http.Response, error)
 
 	return res, err
 }
+
+// RetryMiddleware allows retrying of certain retryable http errors.
+// This only handles very specific status codes, ones that are deemed retryable;
```
Comment (hmlanigan) [inline middleware.go]: s/;/:/

File: middleware.go [hunk 4]
```diff
@@ -131,3 +138,114 @@ func (lr roundTripRecorder) RoundTrip(req *http.Request) (*http.Response, error)
...
+	default:
+		// Don't handle any of the following status codes.
+		return res, false, nil
+	}
+}
+
+// defaultBackoff attempts to workout a good backoff strategy based on the
+// backoff policy or the status code from the response. The
+func (m retryMiddleware) defaultBackoff(resp *http.Response, max time.Duration) time.Duration {
+	if resp.StatusCode == http.StatusServiceUnavailable || resp.StatusCode == http.StatusTooManyRequests {
```
Comment (hmlanigan) [inline middleware.go]: The RFC indicated that only 503 may return Retry-After, doesn't hurt to check others.

File: middleware.go [hunk 5]
```diff
@@ -131,3 +138,114 @@ func (lr roundTripRecorder) RoundTrip(req *http.Request) (*http.Response, error)
...
+		if header := resp.Header.Get("Retry-After"); header != "" {
+			// Attempt to parse the header from the request.
+			retry, err := strconv.ParseInt(header, 10, 64)
```
Comment (hmlanigan) [inline middleware.go]: Found the hard way, the retry value can also be in time.RFC1123 format: https://github.com/go-goose/goose/blob/v2/http/client.go#L384-L409

Comment (SimonRichardson) [reply middleware.go]: NICE, thanks for this. I wonder if we can maybe drop this one in goose eventually as we can pass in a juju/http client? Either way this is great information.

Fixed in e47bd4b

**Comment by SimonRichardson** [comment]
!!build!!

**Comment by SimonRichardson** [comment]
!!build!!

File: middleware.go [hunk 6]
```diff
@@ -235,17 +248,36 @@ func (m retryMiddleware) roundTrip(req *http.Request) (*http.Response, bool, err
...
+		// Check for delay in seconds first, before checkking for a http-date
+		seconds, err := strconv.ParseInt(header, 10, 64)
+		if err == nil {
+			return m.clampBackoff(time.Second * time.Duration(seconds))
+		}
+		// Check for http-date.
+		date, err := time.Parse(time.RFC1123, header)
+		if err == nil {
+			return m.clampBackoff(m.clock.Now().Sub(date))
 		}
```
Comment (hmlanigan) [inline middleware.go:273]: We should log bad header data, that the Retry-After value is not valid, and what the value is.

File: middleware.go [hunk 7]
```diff
@@ -235,17 +248,36 @@ func (m retryMiddleware) roundTrip(req *http.Request) (*http.Response, bool, err
...
+// RFC7231 states that the retry-after header can look like the following:
+//
+//  - Retry-After: <http-date>
+//  - Retry-After: <delay-seconds>
+//
+func (m retryMiddleware) defaultBackoff(resp *http.Response, backoff time.Duration) (time.Duration, error) {
+	if header := resp.Header.Get("Retry-After"); header != "" {
+		// Attempt to parse the header from the request.
+		//
+		// Check for delay in seconds first, before checkking for a http-date
```
Comment (hmlanigan) [inline middleware.go]: s/checkking/checking/

File: middleware.go [hunk 8]
```diff
@@ -207,7 +218,9 @@ func (m retryMiddleware) RoundTrip(req *http.Request) (*http.Response, error) {
 		Attempts: m.policy.Attempts,
 		Delay:    m.policy.Delay,
 		BackoffFunc: func(delay time.Duration, attempts int) time.Duration {
-			return m.defaultBackoff(res, delay)
+			var duration time.Duration
+			duration, backOffErr = m.defaultBackoff(res, delay)
```
Comment (hmlanigan) [inline middleware.go:224]: where is backOffErr evaluated? or logged?

**Review by hmlanigan — APPROVED** [review]
A few minor things to resolve, o.w. LGTM

File: middleware.go (hunk 5 above)
Comment (SimonRichardson) [reply middleware.go]: https://github.com/juju/http/pull/6/files#diff-02f50901140d6057da6310a106670552aa766a093efbc2200fb34c099b762131R200-R202

File: middleware.go (hunk 8 above)
Comment (SimonRichardson) [reply middleware.go:224]: https://github.com/juju/http/pull/6/files#diff-02f50901140d6057da6310a106670552aa766a093efbc2200fb34c099b762131R200-R202

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**